plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -0,0 +1,274 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TypeFamilies     #-}
+{-# LANGUAGE TypeOperators    #-}
+{-# LANGUAGE ViewPatterns     #-}
+{-
+|
+An inlining pass.
+
+This pass is essentially a copy of the PIR inliner, and should be KEPT IN SYNC with it.
+It's hard to do this with true abstraction, so we just have to keep two copies reasonably
+similar.
+
```

Comment:
> We have some confidence in the PIR optimiser because of the types, but they're gone in UPLC; however, if the UPLC optimiser is highly similar to the PIR one, maybe some of the confidence carries over.

Yeah, that's kind of my feeling. It would be even better if they actually shared code, but it's very unclear to me how to do that without trying to construct some very awkward abstraction that covers both PIR terms and UPLC terms...

Good idea about the comment.